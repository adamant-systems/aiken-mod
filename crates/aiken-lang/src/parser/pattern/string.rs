use crate::{
    ast::{ByteArrayFormatPreference, UntypedPattern},
    parser::{error::ParseError, literal, token::Token},
};
use chumsky::prelude::*;

pub fn parser() -> impl Parser<Token, UntypedPattern, Error = ParseError> {
    literal::string().validate(|_, location, emit| {
        emit(ParseError::match_string(location));

        UntypedPattern::ByteArray {
            location,
            value: Vec::new(),
            preferred_format: ByteArrayFormatPreference::Utf8String,
        }
    })
}

#[cfg(test)]
mod tests {
    use crate::assert_expr;

    #[test]
    fn pattern_string() {
        assert_expr!(
            r#"
            when foo is {
              @"foo" -> True
            }
            "#
        );
    }
}
