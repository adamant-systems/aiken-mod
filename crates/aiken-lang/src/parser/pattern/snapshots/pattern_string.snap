---
source: crates/aiken-lang/src/parser/pattern/string.rs
description: "Invalid code (parse error):\n\nwhen foo is {\n  @\"foo\" -> True\n}\n"
---
[
    ParseError {
        kind: PatternMatchOnString,
        span: 16..22,
        while_parsing: None,
        expected: {},
        label: Some(
            "cannot pattern-match on string",
        ),
    },
]
